documentation/correctness.adoc

```diff
@@ -0,0 +1,106 @@
+
+This specification admits some flexibility on the first three, JSON-centric levels.
+That is because the behavior and adaptability of JSON parsers differ from implementation to implementation.
+We simply can't (and don't want to demand) that a LionWeb implementation always produces issues, even if they're recoverable or ignorable anyway.
+
+
+== Overview of the levels
+
+We first give an overview of the various levels, and what each pertains to, at first without specifying the corresponding kinds of issues that may arise in that level.
+The list is roughly in order from low-level to language(s)-specific.
+
+[horizontal]
+
+Low-level syntactic::
+Pertains to the text that's (supposed to be) the JSON serialization chunk of a model.
+
+Structural syntactic::
+Pertains to the syntactic structure of the parsed JSON text.
+It's essentially equivalent to applying a generic JSON Schema.
+
+Hierarchical::
+Pertains to non-local, inter-node constraints on the parsed JSON text.
+An example is that a node that declared a parent node must be contained as a child of that parent node.
+
+Meta-structural::
+This means that the serialization chunk conforms to the language(s) it declares to be an instance of.
+An example is that a node must declare values for exactly the features of the classifier it declares.
+
+Referential::
+This pertains to whether references resolve within the chunk.
+An unresolvable reference doesn't prevent being able to work with the model, but it often prohibits the next level –– Constraints -- from producing meaningful results.
+
+Constraints::
+This pertains to a language(s)-specific notion of correctness.
+An incorrectness would typically either lead to a direct failure in the model's semantics -- i.e.: its execution through interpretation, or generating code and running that --, or to the result of the semantics not making sense in the (context of the) domain.
+This distinction effectively divides this level in two sublevels, but it's up to the language(s) designers to expose those.
+In our experience, a significant part of constraints is "`type-informed`" which means that _type computation/derivation_ is probably an intrinsic part of the constraints aspect of any language.
+
+[NOTE]
+====
+LionWeb uses explicit, ID-based references, which means that scoping is not needed to resolve references.
+Nevertheless, scoping probably still plays some role in any language.
+Consider the following constraint in an ERD-style language: "`in a reference to a field of a referenced entity, the field referenced must be a field of the entity referenced`".
+Even though this is a constraint, it can be interpreted as a scoping rule as well, and the language's UI should take it into account when providing content assist to the language's users.
+====
+
+
+== Overview of the severity
+
+Each issue has a _severity_ that is an indication of the recoverability of the issue.
+They are:
+
+[horizontal]
+
+Ignorable::
+
+Fixable::
```

Comment:
What about calling it `Recoverable`?

Reply:
“Fully recoverable”, then, to distinguish from “partially recoverable”.